framework/generated/dx12_generators/dx12_base_generator.py

```diff
@@ -0,0 +1,662 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2021 LunarG, Inc.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to
+# deal in the Software without restriction, including without limitation the
+# rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+# sell copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+# IN THE SOFTWARE.
+
+import sys
+from base_generator import BaseGenerator, BaseGeneratorOptions, ValueInfo
+
+
+class Dx12GeneratorOptions(BaseGeneratorOptions):
+    """Options for generating C++ function declarations for Dx12 API."""
+
+    def __init__(
+        self,
+        blacklists=None,
+        platform_types=None,
+        filename=None,
+        directory='.',
+        prefix_text='',
+        protect_file=False,
+        protect_feature=True
+    ):
+        BaseGeneratorOptions.__init__(
+            self, blacklists, platform_types, filename, directory, prefix_text,
+            protect_file, protect_feature
+        )
+
+
+class Dx12BaseGenerator(BaseGenerator):
+
+    ARRAY_SIZE_LIST = [
+        ['D3D12_AUTO_BREADCRUMB_NODE', 'pCommandHistory', 'BreadcrumbCount'],
+        ['D3D12_AUTO_BREADCRUMB_NODE1', 'pCommandHistory', 'BreadcrumbCount'],
+        [
+            'D3D12_AUTO_BREADCRUMB_NODE1', 'pBreadcrumbContexts',
+            'BreadcrumbContextsCount'
+        ],
+        [
+            'D3D12_FEATURE_DATA_PROTECTED_RESOURCE_SESSION_TYPES', 'pTypes',
+            'Count'
+        ],
+        [
+            'ID3D12GraphicsCommandList_OMSetRenderTargets',
+            'pRenderTargetDescriptors',
+            'RTsSingleHandleToDescriptorRange ? 1 : NumRenderTargetDescriptors'
```

Comment:
This generates code which is incorrect in the case that NumRenderTargetDescriptors is < 1 such as when the application has called the `ID3D12GraphicsCommandList::OMSetRenderTargets()` method with no render targets specified. In this case (seen in the wild), the code on this line overrides that zero count with a one, which causes a memory copy from a junk address to be initiated.

A more robust line would be something like the uglier:

```
NumRenderTargetDescriptors ? (RTsSingleHandleToDescriptorRange ? 1 : NumRenderTargetDescriptors) : 0
````